crates/circuit/src/register.rs

```diff
@@ -0,0 +1,1054 @@
+                let sequence = py_sequence_index.with_len(self.size())?;
+                match sequence {
+                    crate::slice::SequenceIndex::Int(idx) => {
+                        let Some(bit) = self.0.get(idx) else {
+                            return Err(CircuitError::new_err("register index out of range"));
+                        };
+                        Ok(Box::new(std::iter::once(bit)))
+                    }
+                    _ => {
+                        let result: Vec<BitInfo> = key
+                            .iter_with_size(self.size())?
+                            .map(|idx| -> PyResult<BitInfo> {
+                                self.0
+                                    .get(idx)
+                                    .ok_or(CircuitError::new_err("register index out of range"))
+                            })
+                            .collect::<PyResult<_>>()?;
+                        Ok(Box::new(result.into_iter()))
+                    }
+                }
+            }
+            SliceOrList::List(_) => {
+                let result: Vec<BitInfo> = key
+                    .iter_with_size(self.size())?
+                    .map(|idx| -> PyResult<BitInfo> {
+                        self.0
+                            .get(idx)
+                            .ok_or(CircuitError::new_err("register index out of range"))
+                    })
+                    .collect::<PyResult<_>>()?;
+                Ok(Box::new(result.into_iter()))
+            }
+        }
+    }
+
+    fn iter(&self) -> impl ExactSizeIterator<Item = BitInfo> + '_ {
+        (0..self.size()).map(|bit| self.0.get(bit).unwrap())
+    }
+}
+
+/// Correctly extracts a Slice or a Vec from Python.
+#[derive(FromPyObject)]
+enum SliceOrList<'py> {
+    Slice(PySequenceIndex<'py>),
+    List(Vec<isize>),
+}
+
+impl SliceOrList<'_> {
+    pub fn iter_with_size(&self, size: usize) -> PyResult<Box<dyn Iterator<Item = usize>>> {
+        match self {
+            SliceOrList::Slice(py_sequence_index) => {
+                let sequence = py_sequence_index.with_len(size);
+                match sequence {
+                    Ok(sequence) => Ok(Box::new(sequence.iter())),
+                    Err(e) => Err(e.into()),
+                }
+            }
+            SliceOrList::List(items) => {
+                let items: Vec<usize> = items
+                    .iter()
+                    .copied()
+                    .map(|idx| {
+                        if idx.is_negative() {
+                            size.checked_sub(idx.unsigned_abs())
+                                .ok_or(PySequenceIndexError::OutOfRange.into())
+                        } else {
+                            Ok(idx as usize)
+                        }
+                    })
+                    .collect::<PyResult<_>>()?;
+                Ok(Box::new(items.into_iter()))
+            }
+        }
+    }
+}
```

Comment:
This should probably be in `slice.rs` instead of inline here.

Reply:
This is an extra quirk of `Register`, not standard Python behaviour.  I removed the extra struct in 883d82334 anyway.